Content.Server/GameObjects/EntitySystems/ConstructionSystem.cs

```diff
@@ -27,16 +34,25 @@ internal class ConstructionSystem : Shared.GameObjects.EntitySystems.Constructio
 #pragma warning disable 649
         [Dependency] private readonly IPrototypeManager _prototypeManager;
         [Dependency] private readonly IMapManager _mapManager;
-        [Dependency] private readonly IServerEntityManager _serverEntityManager;
+        [Dependency] private readonly IServerNotifyManager _notifyManager;
 #pragma warning restore 649
 
+        private readonly Dictionary<string, ConstructionPrototype> _craftRecipes = new Dictionary<string, ConstructionPrototype>();
+
         /// <inheritdoc />
         public override void Initialize()
         {
             base.Initialize();
 
+            foreach (var prototype in _prototypeManager.EnumeratePrototypes<ConstructionPrototype>())
+            {
+                _craftRecipes.Add(prototype.Result, prototype);
+            }
+
             SubscribeNetworkEvent<TryStartStructureConstructionMessage>(HandleStartStructureConstruction);
             SubscribeNetworkEvent<TryStartItemConstructionMessage>(HandleStartItemConstruction);
+
+            SubscribeLocalEvent<AfterAttackMessage>(HandleToolInteraction);
```

Comment:
Why does this happen in this way, instead of having an `IInteractUsing` or `IInteractAfter` on the component?

Reply:
Because the construction logic code is inside the `ConstructionSystem`, so there isn't a reason why the `SharedToolComponent` would subscribe to those events with `IInteractAfter`.

> All game logic should exist inside an EntitySystem. EntitySystems should always be isolated/abstracted from each other, and only communicate through the data in a component, tag components, or the EventBus. Components should only contain data, no game logic.

Reply:
I don't know that that design rule is followed in general. I think there's a fair few components with game logic in them.

At the very least, this could have the construction component implement `IInteractUsing` and send a custom message to the system, so that the construction system isn't subscribed to `AfterAttackMessage`. That way the component is the hook into EntitySystem, rather than there being a level of indirection. I would expect click interaction to come from a component, and not a click listener.

Reply:
Putting game logic in systems is a good idea *when it makes sense*. For a case like construction/deconstruction, where many item types with no shared components can be involved in a construction graph, it probably makes sense. For 99% of "regular plain old game code" using interface-on-component is fine. While being purist ECS is nice and all, it's a lot of boilerplate, harder to wrap your head around (code is now split into 2 files at *least*), etc...

Reply:
We had a discussion about this in discord. 700 EntitySystems subscribed to ALL IAfterAttack events, and filtering with TryGetComponent scales poorly, so by default you should be using that behavior interface on a component to raise events. In this specific case this subscription contains handling for multiple component interactions (Material -> Construction, Tool -> arbitrary entity, Tool -> Construction, arbitrary entity -> construction). It is less complex to have this as a single event rather than having the components raise the event, and the support for (arbitrary entity -> construction) interaction is not possible using component events. Therefore, the code should stay as it is.